### DISCUSSION TITLE
[Error - JAVA] - vips_crop calls using JNA result in "no property named `'"

### DISCUSSION THREAD

**Discussion by SnowdingerrDev** [libvips/libvips#4289] [General]
Hello,
I've been looking into how to integrate libvips in my java application,
but I've been encountering issues with few JNA calls.
and while attempting to play with the crop method to only pick single parts of an image

**Describe the issue**
using this as a jna library
```
interface VipsLibrary extends Library;
int vips_crop(Pointer in, Pointer[] out, int left, int top, int width, int height);
```
and then using it as debug in my Libvips training
```
int cropResult = VipsLibrary.INSTANCE.vips_crop(copiedImage[0], croppedImage, 1300, 3400, 1200, 1000);
if (copyResult != 0) System.err.println("Error copying image: " + VipsLibrary.INSTANCE.vips_error_buffer());
```
results in this error being thrown in the console
-> `Error cropping image: extract_area: no property named `'`

I've been digging through the libvips documentation quite a lot, but I've been struggling to find a reliable solution."

**Environment**
- OS : Windows 11
- vips : 8.16
- Jdk11 with maven
- JNA 5.8.0

I've also attempted to use this but still failed with the same message
```
interface Library;
int vips_extract_area(Pointer in, Pointer[] out, int left, int top, int width, int height);
```

Do anyone knows a fix ?
thanks in advance

**Comment by jcupitt** [comment]
Hello @SnowdingerrDev,

You need a NULL at the end of the param list to show there are no extra optional arguments.

However ... I think this is not a good way to use libvips from Java. The libvips C interface is supposed to be compact and simple for humans to write, but it's terrible for language bindings due to varargs and the lack of type safety.

libvips has two C APIs. The `vips_crop()` etc. interface is the one for humans, and that's a very thin skin over the lower-level GObject API. The GObject API is designed for language binding and has no varargs, is typesafe, and supports (almost) full introspection. 

We had a discussion like this one a few months ago and @lopcode went off and made a nice Java binding of the whole of libvips using FFM and the GObject interface:

https://github.com/libvips/libvips/discussions/4112

The binding is here:

https://github.com/lopcode/vips-ffm

I'd have a look at that, it should do what you want (I think).

**Reply by jcupitt** [reply]
For example, here's the code for `vips_crop()`:

https://github.com/libvips/libvips/blob/master/libvips/conversion/extract.c#L299-L328

All it does it gather the arguments and forward them to the libvips GObject interface. It's much safer and simpler to call the GObject layer directly from Java and not go via the C API.

This chapter in the docs has more background:

https://www.libvips.org/API/current/binding.html

**Reply by SnowdingerrDev** [reply]
hello @jcupitt,
thank you for providing this solution,
the issues I had with my libvips implementation were in fact due to the missing NULL at the end of each methods I was using that required an "options" string list.

however as for the vips-ffm you've provided, unfortunatly the project I will be working on once training is on the jdk11 version, but it's still a great recommendation, I'd love to use it in the future !

to anyone trying in the future it simply adds this (which may be me again some times in the future):
```
interface VipsLibrary extends Library;
int vips_crop(Pointer in, Pointer[] out, int left, int top, int width, int height, String... options);
```
and use it as such:
```
int cropResult = VipsLibrary.INSTANCE.vips_crop(copiedImage[0], croppedImage, 1300, 3400, 1200, 1000, null);
```
